Documentation improvement (#101)
lxylxy123456/ngshare: docs/user_guide/notes.rst

Notes for Administrators
========================

> Make sure to completely read and understand the following before putting ``ngshare`` into production

Reviewer: ```suggestion
Make sure to completely read and understand the following before putting ``ngshare`` into production.
```

Race Condition
--------------

> ngshare should NOT run concurrently, or there may be race conditions and data may be corrupted. For example, do not create multiple ngshare instances that share a same underlying database.

Reviewer: ```suggestion
ngshare should NOT be run concurrently, or there may be race conditions and data may be corrupted. For example, do not create multiple ngshare instances that share the same underlying database.
```

Database Upgrade
----------------

> ngshare checks database version every time it starts up. If database version is older than ngshare version, it performs schema and data migration. Normally this may only happen after ngshare is updated.

Reviewer: ```suggestion
ngshare checks the database version every time it starts up. If the database version is older than the ngshare version, it performs schema and data migration. Normally this only happens after ngshare is updated.
```

Author: Abigail and I changed it to "ngshare checks the database version every time it starts up. If the database version is older than the ngshare version, it performs schema and data migration. Normally, this may only happen after ngshare is updated."

I think "may" is necessary because sometimes migration does not happen after ngshare is updated.

Reviewer: The current sentence has two possible meanings:
1. "Under normal circumstances, migration is only allowed after ngshare is updated. Migration can still occur against the rules. Behavior under atypical circumstances is not explained here." In this version, "may" refers to permission, as in "May I come in?".
2. "Under normal circumstances, migration only occurs after ngshare is updated or at some other time. Behavior under atypical circumstances is not explained here." In this version, "may" modifies the word "only", which leads to a contradiction.

None of these meanings seem to match what you are trying to say. I think there are two problems here. "Normally" isn't the right word to use, and "may" modifies the wrong thing. If I understand correctly, migration can only occur due to ngshare being upgraded, but not all upgrades cause a migration. 

Also, I think "this" technically refers to the migration, but some people may misread and think it refers to the entire process of checking the database version and performing a migration if necessary. My suggestion is to change the sentence to something like "Migrations happen only after ngshare is updated, but migrations do not necessarily occur after every update.", or simply "Migrations happen only after ngshare is updated." because the second part is not very important and is implied anyway.

Author: I think here it is better to be more explicit:

> ngshare checks the database version every time it starts up. If the database version is older than the ngshare version, it performs schema and data migration. 
> Under normal circumstances, migrations only happen after ngshare is updated and the update involves change in database structure. The ngshare database update log can be found in :doc:`/contributer_guide/alembic`.

> ngshare checks database version every time it starts up. If database version is older than ngshare version, it performs schema and data migration. Normally this may only happen after ngshare is updated.

Reviewer 2: ```suggestion
ngshare checks database version every time it starts up. If database version is older than ngshare version, it performs schema and data migration. Normally, this may only happen after ngshare is updated.
```

The check can be disabled using ``--no-upgrade-db`` but do not disable it unless you have a good reason and know the possible consequences. 

The database update history can be found in :doc:`/contributer_guide/alembic`.

Database Backup
---------------

> ngshare users should regularly backup the database. Especially the database should be backed up before updating ngshare. Because after the update, the schema and data migration may corrupt the database.

Reviewer: ```suggestion
ngshare users should regularly back up the database, especially before updating ngshare. The schema and data migration may corrupt the database.
```

Author: @aalmanza1998 @Lawrence37 I am changed it to (2 paragraphs):

ngshare users should regularly back up the database in case of corruption.

The database should be backed up before updating ngshare because the schema and data migration may corrupt the database.

> ngshare users should regularly backup the database. Especially the database should be backed up before updating ngshare. Because after the update, the schema and data migration may corrupt the database.

Reviewer 2: ```suggestion
ngshare users should regularly backup the database. The database should be backed up before updating ngshare because the schema and data migration may corrupt the database.
```

Reviewer 2: please let me know if I misunderstood something here

Database Upgrade
----------------
ngshare checks the database version every time it starts up. If the database version is older than the ngshare version, it performs schema and data migration. 

> Under normal circumstances, migrations only happen after ngshare is updated and the update involves change in database structure. The ngshare database update log can be found in :doc:`/contributer_guide/alembic`.

Reviewer: ```suggestion
Under normal circumstances, migrations only happen after ngshare is updated and the update involves a change in database structure. The ngshare database update log can be found in :doc:`/contributer_guide/alembic`.
```

Author: I think "changing database structure" is better.

Reviewer: "involves changing the database structure"

Author: Fixed
